src/step-function.ts

```diff
@@ -466,6 +466,428 @@ export namespace $SFN {
+      if (!data || !algorithm) {
+        throw new SynthError(
+          ErrorCodes.Invalid_Input,
+          "Expected Hash data and algorithm arguments to be provided."
+        );
+      }
+
+      return context.evalContext(
+        call,
+        ({ evalExprToJsonPath, evalExprToJsonPathOrLiteral }) => {
+          const dataOut = evalExprToJsonPath(data.expr);
+          const algorithmOut = evalExprToJsonPathOrLiteral(algorithm.expr);
+
+          if (
+            ASLGraph.isLiteralValue(algorithmOut) &&
+            typeof algorithmOut.value !== "string"
+          ) {
+            throw new SynthError(
+              ErrorCodes.Invalid_Input,
+              "Expected Hash algorithm argument to be a string or reference."
+            );
+          }
+          const temp = context.newHeapVariable();
+          return {
+            Type: "Pass",
+            Parameters: {
+              "out.$": `States.Hash(${dataOut.jsonPath}, ${
+                ASLGraph.isJsonPath(algorithmOut)
+                  ? algorithmOut.jsonPath
+                  : `'${algorithmOut.value}'`
+              })`,
+            },
+            ResultPath: temp,
+            Next: ASLGraph.DeferNext,
+            output: {
+              jsonPath: `${temp}.out`,
+            },
+          };
+        }
+      );
+    },
+  });
+
+  /**
+   * Use the States.MathRandom intrinsic function to return a random number between the specified start and end number.
+   * For example, you can use this function to distribute a specific task between two or more resources.
+   *
+   * @param start - starting number, must be an integer.
+   * @param end - ending number, must be an integer.
+   * @param seed
+   */
+  export const random = makeStepFunctionIntegration<
+    "SFN.Random",
+    (start: number, end: number, seed?: number) => number
+  >("SFN.Random", {
+    asl(call, context) {
+      const [start, end, seed] = call.args;
+
+      if (!start || !end) {
+        throw new SynthError(
+          ErrorCodes.Invalid_Input,
+          "Expected Random start and end arguments to be provided."
+        );
+      }
+
+      return context.evalContext(call, ({ evalExprToJsonPathOrLiteral }) => {
+        const startOut = evalExprToJsonPathOrLiteral(start.expr);
+        const endOut = evalExprToJsonPathOrLiteral(end.expr);
+        const seedOut = seed
+          ? evalExprToJsonPathOrLiteral(seed.expr)
+          : undefined;
+
+        if (
+          ASLGraph.isLiteralValue(startOut) &&
+          typeof startOut.value !== "number"
+        ) {
+          throw new SynthError(
+            ErrorCodes.Invalid_Input,
+            "Expected Random argument to be a number or reference."
+          );
+        } else if (
+          ASLGraph.isLiteralValue(endOut) &&
+          typeof endOut.value !== "number"
+        ) {
+          throw new SynthError(
+            ErrorCodes.Invalid_Input,
+            "Expected Random argument to be a number or reference."
+          );
```

Comment:
Let's create assertion functions to try and reduce some of the redundancy in these.